### DISCUSSION TITLE
Naming conventions

### DISCUSSION THREAD

**Discussion by nsmith-** [scikit-hep/vector#17] [General]
Hi, not sure where to put this but I wanted to check that the work I did for coffea vectors will transfer seamlessly to this when it is ready. I have the following [classes](https://coffeateam.github.io/coffea/modules/coffea.nanoevents.methods.vector.html) (implementation [here](https://github.com/CoffeaTeam/coffea/blob/master/coffea/nanoevents/methods/vector.py)) and was wondering if you could review the naming convention and see if it matches the plans for this library, at least for the member attributes. I see that the ROOT TVectors prefer to avoid using `r` and `rho` for ambiguity's sake since depending where you look you find `r` for the 2D polar radius and `rho` for the 3D distance and vice-versa (e.g. [TVector](https://root.cern.ch/doc/master/classTVector3.html) vs. https://en.wikipedia.org/wiki/Spherical_coordinate_system)
In addition, I wonder if the inheritance approach will be taken for this library or not (I'm not sure myself if it makes sense) and if there are also plans for Point and "Point with origin" classes?

**Comment by nsmith-** [comment]
P.S. I just pushed a much more thoroughly documented module to help with this.

P. P. S. I would happily make a PR with the contents of that module so non-coffea users can have something to get started with on awkward1. I feel like awkward without vector mixins is like cake without ice cream

**Comment by donalhill** [comment]
Hi all - I just wanted to chip in here to say that some FCC studies that I'm starting [here](https://github.com/donalrinho/fcc_python_tools) with uproot and awkward will benefit a lot from vector. I anticipate a good amount of uptake from people just starting FCC studies, so it will be great to have vector as part of the setup. 

On this note - it may be that Coffea can help us too, rather than me reinventing the wheel! I'm coming to this work from a flavour physics perspective, so building lots of exclusive decay modes and studying the momentum, impact parameter, lifetime resolutions e.t.c.

**Comment by henryiii** [comment]
I finally released Hist 2.0, so this will be my next (large) priority. I have a talk to write for Wednesday, so likely after that? Help and example appreciated!

**Comment by eduardo-rodrigues** [comment]
For reference - when discussing naming conventions, hence the API, we should not forget all that we're done in the past and is still available in the `scikit-hep` metapackage, see `https://github.com/scikit-hep/scikit-hep/tree/master/skhep/math` . I will be involved in the discussions for sure :-).

**Comment by henryiii** [comment]
Let's start by discussing `.dot`. Should we:
1. Provide only one form. `v1.dot(v2)`.
2. Provide an alternate spelling, `v1 @ v2`.
3. Provide an alternate spelling, `v1 * v2`.

For now, I've implemented 3, but I might rather 1 or 2. Thoughts? Please reply inline, new discussion comments should be about different aspects of the API.

**Reply by jpivarski** [reply]
Personally, I would do (1) _and_ (2) _and not_ (3).

   1. I think it's good for every operation to have a `[A-Za-z_][A-Za-z_0-9]*` spelling, at least as an alternative. Reasons for it sound far-fetched until you actually run into them—perhaps someone wants to duplicate your interface in another language someday, but `@` is not allowed in that language, or whatever. Actually, here's a more likely one: a method name is more discoverable by tab-completion than a binary operator. Here's another one: you want to ensure the correct precedence without having to look up Python's operator precedence and without the code starting to look like LISP. Here's another (maybe not relevant anymore): Python 2, which doesn't have a `@`. Anyway, `dot` is a very good name for this method. It's easy to guess what it means.
   2. That said, the dot product is symmetric and needed frequently, at least as frequently as operations like `+` that get their own operators. Python introduced `@` specifically for this. Using an operator means that one of the two vectors doesn't have to be in parentheses, which is good to express that symmetry. The reasons for dot to be implemented as `@` are stronger than any others (including (1)).
   3. If there's one operator that provides this function, there shouldn't be two. It would be a nice safety feature if `*` were used exclusively for scalar multiplication and `@` were used exclusively for dot product and matrix transformation. I just checked and NumPy agrees with me:

```python
>>> np.array([1, 2, 3]) * 10
array([10, 20, 30])

>>> np.array([1, 2, 3]) @ 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)

>>> np.array([[1, 2, 3]]) @ 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)

>>> np.array([[1], [2], [3]]) @ 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

It's a nice way to make sure that someone doesn't dot-product something when they meant to scalar multiply.

Now the only sad thing is that Python doesn't have an operator for cross product. There isn't a good ASCII character left for that: `^`? `$`? Yuck. For something like this, I wouldn't be opposed to Python adopting Unicode—memorizing a small number of frequently used code-points is not bad at all: Ub7 (·) and Ud7 (×). The control-shift-U keystroke for entering this mode is pretty universal these days. Or maybe a space-separated `x`, which I think is not currently allowed in Python's syntax, but it would have to be very carefully reviewed to be sure that it doesn't lead to bad consequences. Anyway, that's way beyond the scope of this question.

**Reply by eduardo-rodrigues** [reply]
I take @jpivarski's points. I would also have (1) for sure (that's a requirement since expected by pretty much anyone/everyone) and (2) seems a nice addition. Yes, we will have 2 ways but ...

Going along NumPy seems a good move as well.

**Reply by henryiii** [reply]
Option 2 and 3 were both supposed to include 1 already. Though I'm really not against writing `v1.__mat__(v2)` - if you don't like writing that, you shouldn't be using Python 2. ;)

What about addition with a scalar, addition with a vector, and multiplication with a scalar? Are those supposed to have alternate spellings too?

**Reply by jpivarski** [reply]
At first, I would say, "Yes!" In analysis code, I'd usually opt for the binary operator, but in formal programming, I do prefer words for clarity and uniformity. For instance, I definitely replace `&` and `|` with `numpy.bitwise_and` and `numpy.bitwise_or` (or the `logical_and` and `logical_or`, since they are different and sometimes the difference is important). I did exactly this on a pull request I received a few months ago. As single-character operators, the operation was too easy to miss seeing.

Having `add` and `multiply` (which are the spellings of the NumPy ufuncs) as vector methods would make a lot of sense, allowing an expression in formal code to uniformly consist of method applications. The direction of operator precedence would then be strictly left-to-right, which also helps. There wouldn't be a distinct `add` for the scalar and vector cases: it is conventional to broadcast. (Dot product, by contrast, isn't broadcasting.)

However, NumPy doesn't do that. NumPy has a `dot` method that is equivalent to `@`, and of course there are `numpy.add` and `numpy.multiply` ufuncs, but not methods. That breaks pattern with a lot of other methods that are also free-standing functions, but the NumPy project is trying to cut down on that duplication anyway. So maybe there's enough of a precedent for addition and scalar multiplication to be done with binary operators that `add` and `multiply` methods don't need to exist.

I'd say that's your call: consistently adding an English method name for every operation or following NumPy in doing `dot` but not `add` and `multiply`.

**Reply by henryiii** [reply]
I'd like to go with avoiding duplication. I don't see anything wrong with calling `v1.__add__(v2)` if someone wants to be explicit and/or avoid precedence issues. It's a little long with the dunder's, but it's clear and fully mandated by Python rather than having to look up whether a method is called `mul` or `multiply` or `multiply_scalar`. `dot` is special because you are actually doing a dot product, not a matrix multiplication or element-wise multiplication, so only providing something like `__mat__` is unintuitive.

This might be partially why NumPy does not do this - why provide `.add` when `.__add__` is identical and the idiomatic way to spell it in Python?

**Reply by jpivarski** [reply]
Even though the special methods (`__add__` and such) have a meaning that is guaranteed by Python, I think they fall under the "not public API; do not use" rule and therefore I don't use them in formal code.

Anyway, I don't have strong feelings on adding the `add` and `multiply` methods, particularly because NumPy doesn't, and that's an easy convention-setting rule to apply.

The `dot` is special for an additional reason: it should have symmetry with `cross` and there's no Python operator for a vector cross-product.

**Reply by nsmith-** [reply]
I'm against the proposition that dunder methods (`__add__`) fill the role of an instance method here, as it appears private. For what its worth, I did include `v.add` and `v.multiply`, keeping the naming suggested by the equivalent numpy ufunc that the binary method would call (e.g. https://github.com/CoffeaTeam/coffea/blob/master/coffea/nanoevents/methods/vector.py#L112 ) but personally, I wouldn't be against having a vector namespace with free functions that apply pairwise.

**Reply by henryiii** [reply]
Dunder methods are definitely not private - you can't access the `__name__` of a class any other way (AFAIK). And I always use `f"{self.__class__.__name__}()"` instead of `f"{type(self).__name__}()"` - I prefer the fewer parenthesis, and `"{self.__class__.__name__}()".format(self=self)` is supported, while you can't do that inline with type. There's also `super().__init__()`, etc. Python literally treats these different from double-underscore but not ending in double underscore names, which are private (name mangled outside the class, for inheritance purposes mostly). Double underscore just means they are reserved for a specific meaning - add in this case. Which is the correct meaning.

My point was that if you do need to call a method, it's quite safe to call `__add__`, and adding a `.add` is entirely optional - and most languages support overloading + anyway, which is a more natural way to write it. I really want to keep duplication down in the core library, since we already have a _ton_ of duplication anyway, for the initial five backends.

@mayou36 might have a strong opinion, though - the TensorFlow method of writing things may tend to look like `LorentzZYXT.__add__(v1, v2)` / `LorentzZYXT.add(v1, v2)` - I'm mildly okay with the former, and I don't like having to learn something library specific (`add`) vs. something standard (`__add__`), but it does look odd if that's the main way to use it.

> naming suggested by the equivalent numpy ufunc that the binary method would call

In NumPy, this is called `np.add` or `array.__add__`, there is no `array.add`.

> vector namespace with free functions that apply pairwise.

In Python 3, you can use the class methods as free functions, so that's how these will be used if you want free functions. There will be no duplicated namespace like NumPy has. We have far too much duplication as it is. :)

**Comment by henryiii** [comment]
For these discussions, the current Protocol in [src/vector/protocols/lorentz.py](https://github.com/scikit-hep/vector/blob/main/src/vector/protocols/lorentz.py), and you can also look at the parametrized tests in [tests/test_simple_param.py](https://github.com/scikit-hep/vector/blob/05d1a1dcbe59f0443e0e6bc875b2f10c6801912c/tests/test_simple_param.py) (static link, will likely rename soon).
